Add Entity Relationship Diagram and Attribute Table to Documentation (#6)
mojaloop/auth-service: docs/README.md

# auth-service/Docs

Documentation for the auth service
## Database

The following Figure depicts the Entity Relationship Diagram of the `Consents` table for the authorization service:

<p align="center">
  <img src="ErDiagram.png">
</p>

The following table depicts the types of each attribute in our table, algonside their constraints:

|Attribute| Represented Internal Type | Constraints|
|:---------:|:---------------------------:|:------------:|

> |`ID`|BIGINT|Unique - Primary Key|

Reviewer: A couple notes to bring this into line with other Mojaloop standards:
- `ID` should be `id`, and of type `VARCHAR(36)` 
  - take a look at [central-ledger](https://github.com/mojaloop/central-ledger/blob/master/migrations/300200_transfer.js#L32) for example (but don't copy the `tableNameId` pattern, `id` is much better)
  - Also, just for reference: https://github.com/mojaloop/mojaloop-specification/blob/master/Supporting%20Files/fspiop-rest-v1.0-OpenAPI-implementation.yaml#L3061

- Tablenames in the database should be snake case: e.g. `initiator_id`, `participant_id`
  - Knex can map from `snake_case` to `camelCase` for us
- `initiator_id` and `participant_id` should probably also be `VARCHAR(32)`, according to the [swagger spec](https://github.com/mojaloop/mojaloop-specification/blob/master/Supporting%20Files/fspiop-rest-v1.0-OpenAPI-implementation.yaml#L3282)

- Can you go into more detail (perhaps in a different table) about the format of `scopes` and `credential`? I thought that `credential` would be expanded into separate, _nullable_ columns based on our conversation the other day, eg:
```
credential_id:
credential_status: 
credential_type:
credential_challenge:
credential_payload:
```
Alternatively, this could also just be a _nullable_ JSON field as well.

- Finally (and I'm really nitpicking here) the table should probably be called `consent`, and not `Consents`

Author: Addressed, and modified, the PR.

# auth-service/Docs

Documentation for the auth service
## Database

The following Figure depicts the Entity Relationship Diagram of the auth-service's schema:

<p align="center">
  <img src="http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/spikerheado1234/auth-service/master/docs/ErDiagram.puml">
</p>

The following table depicts the types of each attribute in the `Consent` table, algonside their constraints:

|Attribute| Represented Internal Type | Constraints|
|:---------:|:---------------------------:|:------------:|
|`id`|VARCHAR(36)|Unique - Primary Key|
|`initiator_id`|VARCHAR(36)|Non-Null|
|`participant_id`|VARCHAR(36)|Non-Null|

> |`credential_id`|TEXT|(Can be Null)|

Reviewer 2: Can we double check on the format of this? I suspect that this would be numeric, but we should probably limit this rather than leave it as an unbounded text field

> |`credential_type`|TEXT|(Can be Null)|

Reviewer 2: This is determined by us and should be limited rather than an arbitrary text field. Maybe `VARCHAR(32)` ?

> |`credential_status`|TEXT|(Can be Null)|

Reviewer 2: The status should be a set of specific values, so definitely not unbounded text. Maybe `VARCHAR(16)`?

Reviewer 3: The status is either `PENDING` or `ACTIVE`. So `VARCHAR(10)` should be good? 
Updated the PR.

> |`initiator_id`|VARCHAR(36)|Non-Null|
> |`participant_id`|VARCHAR(36)|Non-Null|
> |`credential_id`|TEXT|(Can be Null)|
> |`credential_type`|TEXT|(Can be Null)|
> |`credential_status`|TEXT|(Can be Null)|
> |`credential_payload`|TEXT|(Can be Null)|
> |`credential_challenge`|TEXT|(Can be Null)|

Reviewer 2: ```suggestion
|`initiator_id`|`VARCHAR(36)`|Non-Null|
|`participant_id`|`VARCHAR(36)`|Non-Null|
|`credential_id`|`TEXT`|(Can be Null)|
|`credential_type`|`TEXT`|(Can be Null)|
|`credential_status`|`TEXT`|(Can be Null)|
|`credential_payload`|`TEXT`|(Can be Null)|
|`credential_challenge`|`TEXT`|(Can be Null)|
```

Reviewer 2: Just adding code ticks around these.

Reviewer 3: Formatted the data types and updated the PR.

> |`credential_challenge`|TEXT|(Can be Null)|

Reviewer 2: `credential_challenge` is something we can determine the size of. Why not decide on how many bytes we care about and then just go with that? So `CHAR(128)` or something like that?

The following table depicts the types of each attribute in the `Scope` table:

|Attribute| Represented Internal Type | Constraints|
|:---------:|:---------------------------:|:------------:|
|`id`|VARCHAR(36)|Unique - Primary Key|

> |`consent_id`|VARCHAR(36)|Foreign Key referring to the `id` in the consent table|

Reviewer 2: ```suggestion
|`consent_id`|`VARCHAR(36)`|Foreign Key referring to the `id` in the consent table|
```

> |`scope_account_type`|VARCHAR(36)|Non-Null|

Reviewer 2: ```suggestion
|`action`|`VARCHAR(36)`|Non-Null|
```

Reviewer 2: This isn't really an account type, it's an action permitted on an account (e.g., `get balance` or `send money`)

Reviewer 3: Modified data types and updated the PR.

> |`scope_account_number`|VARCHAR(36)|(Can be Null)|

Reviewer 2: ```suggestion
|`account_number`|`VARCHAR(36)`|(Can be Null)|
```

> The reason attributes pertaining to `credentials` and `scope_account_number` can be null is because of the communication sequence established in between the `PISP` and the `authorisation-service`. Where, at the point of first contact with the authorisation-service, these fields are not present. They are available during later parts in the sequence of communication in between the two services.

Reviewer 2: ```suggestion
The reason attributes pertaining to `credentials` and `account_id` can be null is because of the communication sequence established in between the PISP and the auth service. At the point of first contact with the auth service, these fields are not present and instead are added later on in the sequence of communication in between the two services.
```

Reviewer 3: Isn't the credential info populated by the Auth service itself? If my understanding of the sequence diagrams is correct, the service itself creates and populates all the credential attributes, except the public key (payload), when it receives the request from the switch. Thus, the only credential attribute that should be allowed to be NULL is the payload.

The scopes and the account numbers are received by the Auth service in the incoming request too. If there are no scopes, we don't need to add any row in the Scopes table. Hence, they can be Non-Nullable too.

Author: Yes, I agree. The only nullable value should theoretically be `challenge_payload`.

# auth-service/Docs

Documentation for the auth service
## Database

The following Figure depicts the Entity Relationship Diagram of the auth-service's schema:

<p align="center">
  <img src="http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/spikerheado1234/auth-service/master/docs/ErDiagram.puml">
</p>

The following table depicts the types of each attribute in the `Consent` table along with the constraints:

|Attribute| Represented Internal Type | Constraints| Description |
|:---------:|:---------------------------:|:------------:|:---------------------------:|
|`id`|`VARCHAR(36)`|Unique - Primary Key|Alphanumeric consent ID|
|`initiator_id`|`VARCHAR(32)`|Non-Null|Alphanumeric PISP ID|
|`participant_id`|`VARCHAR(32)`|Non-Null|Alphanumeric DFSP ID|

> |`credential_id`|`INT UNSIGNED`|Nullable|Unsigned numeric ID - upto 2^32 records|

Reviewer 2: https://www.w3.org/TR/credential-management-1/#credential says webauthn uses a string

android apis seem to use a string also https://developers.google.com/android/reference/com/google/android/gms/fido/fido2/api/common/PublicKeyCredential#getId()

> |`credential_id`|`INT UNSIGNED`|Unique - Non-Null|Unsigned numeric ID - upto 2^32 records|

Reviewer: I think this **needs** to be _nullable_, since it will be filled in by the PISP - it's the unique identifier of the FIDO key on the device, so won't be filled in until the PISP has signed the challenge.

Reviewer 3: Is there a specific data type for the PISP credential ID or can the auth service specify the type of ID it wants?

|`credential_type`|`VARCHAR(16)`|Non-Null|Alphanumeric enum value|
|`credential_status`|`VARCHAR(10)`|Non-Null|String - `PENDING`/`ACTIVE`|
|`credential_payload`|`TEXT`|Can be Null|Public key string|

> |`credential_challenge`|`CHAR(128)`|Non-Null|Base 64 encoded challenge string|

Reviewer: A _Non-Null_ constraint here implies that the `credential_challenge` is created when the `Consent` object is created. 

However, in [1.5-grant-consent](https://github.com/jgeewax/pisp/blob/master/docs/linking/README.md#15-grant-consent), the credential object is still null, and the challenge isn't created until `CRED-5` in [1.6.1 Requesting a challenge](https://github.com/jgeewax/pisp/blob/master/docs/linking/README.md#161-requesting-a-challenge)

The following table depicts the types of each attribute in the `Scope` table:

|Attribute| Represented Internal Type | Constraints| Description |
|:---------:|:---------------------------:|:------------:|:---------------------------:|
|`id`|`VARCHAR(36)`|Unique - Primary Key|Alphanumeric Scope ID|
|`consent_id`|`VARCHAR(36)`|Non-Null - Foreign Key|Foreign Key for `id` in the `Consent` table|
|`action`|`VARCHAR(36)`|Non-Null|Account scope (enum) action allowed with corresponding consent|

> |`account_number`|`VARCHAR(36)`|Non-Null|Account number for associated scope|

Reviewer: I think this should be `account_id`, and not account number. The `account_id` is an opaque identifier provided by the DFSP, and calling it 'number' is a little too opinionated, in my opinion ;)

# auth-service/Docs

Documentation for the auth service
## Database

The following Figure depicts the Entity Relationship Diagram of the auth-service's schema:

<p align="center">
  <img src="http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/spikerheado1234/auth-service/master/docs/ErDiagram.puml">
</p>

The following table depicts the types of each attribute in the `Consent` table along with the constraints:

|Attribute| Represented Internal Type | Constraints| Description |
|:---------:|:---------------------------:|:------------:|:---------------------------:|
|`id`|`VARCHAR(36)`|Unique - Primary Key|Alphanumeric consent ID|
|`initiator_id`|`VARCHAR(32)`|Non-Null|Alphanumeric PISP ID|
|`participant_id`|`VARCHAR(32)`|Non-Null|Alphanumeric DFSP ID|
|`created_at`|`TIMESTAMP`|Non-Null|Timestamp of consent creation|

> |`credential_id`|`CHAR(256)`|Nullable|Unsigned numeric ID - upto 2^32 records|

Reviewer 2: https://codelabs.developers.google.com/codelabs/fido2-for-android/#5 also shows a string being used.

```suggestion
|`credential_id`|`VARCHAR(256)`|Nullable|Unsigned numeric ID - upto 2^32 records|
```
